doc/design/scope.md

```diff
@@ -0,0 +1,58 @@
+# Scope
+
+### Define
+
+Scope is a context to manage Variables. It mainly contains a map from Variable name to Variable. Net will get and update variable throw scope.
+
+```cpp
+class Variable;
+using VariablePtr = std::shared_ptr<Variable>;
+
+class Scope final {
+public:
+  Scope();
+  Scope(const std::shared_ptr<Scope>& parent);
+
+  //! Get Variable in this scope.
+  //! @return nullptr if no such variable.
+  const VariablePtr& getVar(const std::string& name) const;
+
+  //! Create or get a variable in this scope.
+  VariablePtr& createOrGetVar(const std::string& name);
+
+private:
+  /// variable name -> variable
+  std::unordered_map<std::string, VariablePtr> vars_;
```

Comment:
yangqing has metioned why he prefer `map` instead `unordered_map`

[>//](https://github.com/caffe2/caffe2/blob/v0.7.0/caffe2/core/common.h#L44) Note(Yangqing): NVCC does not play well with unordered_map on some platforms,
>// forcing us to use std::map instead of unordered_map.